FR #956: Provider-calculated split cost allocation (#1394)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/allocatedmethodid.md

# Allocated Method ID

> Allocated Method ID is the identifier for the method defined by the provider which was used for the provider-calculated split cost allocation.

Reviewer: What is an "allocated method"? This isn't clear.

Reviewer 2: How does

> Allocated Method ID is the identifier for the method or calculation defined and documented by the provider which was used to calculate the split cost allocation. This identifier can be used as a search string to find more information specific to this allocation method in the provider's documentation.

sound?

Author: I added a link to [Provider-Calculated Split Cost Allocation](#providercalculatedsplitcosthandling) where the requirement for documentation is included.

https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/2f7b440e3208e5ad12341936c94b200015420f46/specification/attributes/provider_calculated_split_costallocation_handling.md?plain=1#L18

Added a definition for [Allocated Method](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/38e4ee2ba498f45986ff8a9ce3bdddd920051220/specification/glossary.md?plain=1#L11) to the glossary.

Incorporated some of Q's verbiage.

The AllocatedMethodId column adheres to the following requirements:

> * AllocatedMethodId MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports provider-calculated split cost allocation.

Reviewer 3: @kk09v @qquigley 
In the current state, this PR requires `AllocatedMethodId` but does NOT require `AllocatedMethodDetails` if the provider supports split cost allocation.  Should they not be joined at the hip?  If so, should this change to SHOULD?

Reviewer 2: The thinking behind this was that if there is an allocation, then there MUST be documentation about the method somewhere. `AllocatedMethodId` would be used to find that documentation. I personally would love to require `AllocatedMethodDetails`, but we compromised on where it is now partly due to provider pushback, and partly to this being more of an advanced feature thing. iirc, a provider said there might be occasions where they wouldn't be able to provide those details.

But as I type this out, a provider should, at the very least, be able to provide a ratio, which is the only requirement of the `AllocatedMethodDetails` column.

So to answer your question, I think it should stay MUST. But I agree that they're attached at the hip and we've made `AllocatedMethodDetails` flexible enough at this point, that we could make that a MUST as well.

Reviewer 3: Based on the discourse, I suspect it too challenging to make `AllocatedMethodDetails` a MUST and get consensus.

What's an example of what they might put in here?  It would be helpful for my mental model.

Reviewer 2: `AllocatedMethodId ` could be a doc ID maybe like `EKS1234`. Then you could go to the AWS docs and search for `EKS1234` and find documentation the method used for this allocation. Or it could be a title like "Standard EKS Allocation Method', again, something that can be used to search docs to find the math behind the method used.

And then the practitioner could find a doc like this: https://docs.aws.amazon.com/cur/latest/userguide/example-split-cost-allocation-data.html

That works for AWS EKS today. But let's say in the future they adjust their formula. Then `AllocatedMethodId ` would be `EKS1235`. Now we know the formula has changed and is distinct from previously ingested data.

Reviewer 3: If that's the case, then I suggest a change to a different normative requirement.  Standby.

Reviewer 3: See [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1394/files/80d4da7ddf283ead32c551efe6fb8cc2c14e365a#r2363783536).  What do you think?

Reviewer 3: @qquigley accepted my suggestion on the other thread.  Resolving this one!

* AllocatedMethodId MUST be of type String.
* AllocatedMethodId MUST conform to [StringHandling](#stringhandling) requirements.
* AllocatedMethodId nullability is defined as follows:
  * AllocatedMethodId MUST be null when a charge is not related to a provider-calculated split cost allocation.

>   * AllocatedMethodId MUST NOT be null when a charge is related to a provider-calculated split cost allocation.

Reviewer 4: nit:

```suggestion
  * AllocatedMethodId MUST be null when a [*charge*](#glossary:charge) is not related to a provider-calculated split cost allocation.
  * AllocatedMethodId MUST NOT be null when a *charge* is related to a provider-calculated split cost allocation.
```

> * AllocatedMethodId MUST uniquely identify the method used to calculate the split cost allocation in the provider's documentation.

Reviewer 3: Per [this discussion](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1394/files/80d4da7ddf283ead32c551efe6fb8cc2c14e365a#r2360969537), I suggest:

```suggestion
* Provider documentation of a split cost allocation method MUST make reference to a single AllocatedMethodId value.
```

The way it's written, the requirement seems more associated with uniqueness than it does about documentation.

If that doesn't work for you, I'd love for you to propose a revision.

Reviewer 2: I like it. Gonna noodle on this a bit but will probably accept your commit.

Noodling because I want to take a closer look at this whole column to see if any more slight tweaks like this are needed elsewhere in the column definition.

Reviewer 3: Please do noodle!  Definitely don't accept my suggestion blindly, and feel free to consult with your co-author before committing.

## Column ID

AllocatedMethodId

## Display Name

Allocated Method ID

## Description

> Allocated Method ID is the unique identifier for the method defined by the provider which was used to calculate the provider-calculated split cost allocation.

Reviewer 3: ```suggestion
A unique identifier defining the method of provider-calculated split cost allocation.
```
